src/xenia/base/threading_posix.cc

```diff
@@ -47,265 +95,746 @@ void MaybeYield() {
-    assert_always();
-    return false;
+    // Create timer
+    sigevent sev{};
+    sev.sigev_notify = SIGEV_SIGNAL;
+    sev.sigev_signo = GetSystemSignal(SignalType::kHighResolutionTimer);
+    sev.sigev_value.sival_ptr = (void*)&callback_;
+    if (timer_create(CLOCK_REALTIME, &sev, &timer_) == -1) return false;
+
+    // Start timer
+    itimerspec its{};
+    its.it_value = DurationToTimeSpec(period);
+    its.it_interval = its.it_value;
+    return timer_settime(timer_, 0, &its, nullptr) != -1;
   }
 
  private:
   std::function<void()> callback_;
+  timer_t timer_;
 };
 
 std::unique_ptr<HighResolutionTimer> HighResolutionTimer::CreateRepeating(
     std::chrono::milliseconds period, std::function<void()> callback) {
+  install_signal_handler(SignalType::kHighResolutionTimer);
   auto timer = std::make_unique<PosixHighResolutionTimer>(std::move(callback));
   if (!timer->Initialize(period)) {
     return nullptr;
   }
   return std::unique_ptr<HighResolutionTimer>(timer.release());
 }
 
-// TODO(dougvj) There really is no native POSIX handle for a single wait/signal
-// construct pthreads is at a lower level with more handles for such a mechanism
-// This simple wrapper class could function as our handle, but probably needs
-// some more functionality
-class PosixCondition {
+class PosixConditionBase {
  public:
-  PosixCondition() : signal_(false) {
-    pthread_mutex_init(&mutex_, NULL);
-    pthread_cond_init(&cond_, NULL);
+  virtual bool Signal() = 0;
+
+  WaitResult Wait(std::chrono::milliseconds timeout) {
+    bool executed;
+    auto predicate = [this] { return this->signaled(); };
+    auto lock = std::unique_lock<std::mutex>(mutex_);
+    if (predicate()) {
+      executed = true;
+    } else {
+      if (timeout == std::chrono::milliseconds::max()) {
+        cond_.wait(lock, predicate);
+        executed = true;  // Did not time out;
+      } else {
+        executed = cond_.wait_for(lock, timeout, predicate);
+      }
+    }
+    if (executed) {
+      post_execution();
+      return WaitResult::kSuccess;
+    } else {
+      return WaitResult::kTimeout;
+    }
   }
 
-  ~PosixCondition() {
-    pthread_mutex_destroy(&mutex_);
-    pthread_cond_destroy(&cond_);
+  static std::pair<WaitResult, size_t> WaitMultiple(
+      std::vector<PosixConditionBase*>&& handles, bool wait_all,
+      std::chrono::milliseconds timeout) {
+    using iter_t = std::vector<PosixConditionBase*>::const_iterator;
+    bool executed;
+    auto predicate = [](auto h) { return h->signaled(); };
+
+    // Construct a condition for all or any depending on wait_all
+    auto operation = wait_all ? std::all_of<iter_t, decltype(predicate)>
+                              : std::any_of<iter_t, decltype(predicate)>;
+    auto aggregate = [&handles, operation, predicate] {
+      return operation(handles.cbegin(), handles.cend(), predicate);
+    };
+
+    std::unique_lock<std::mutex> lock(PosixConditionBase::mutex_);
```

Comment:
This will probably deadlock if the thread is suspended between locking and waiting.

Reply:
Suspension needs rework to only occur immediately when the thread is in guest code, but deferred until the return to the guest if it happens during a kernel call.

This should be somewhat easy to implement on POSIX because the thread itself will reply to the suspension signal — it can set a flag in its context structure when it enters a kernel function, and the signal handler can check whether to sleep immediately or to defer the suspension.

On Windows, however (while there is a native WaitForMultipleObjects implementation that will work fine, having control over suspension would allow us more to have more granular kernel mutexes rather than just a single global_critical_region which is our current way of preventing suspension when kernel state is being modified, with easier usage of condition variables in the kernel — would possibly be useful for GPU readback in the access violation handler), the target thread can't provide its own response to suspension, and there are no signals. A possible solution is making suspension synchronous (which is contrary to the way it actually works on Windows, but possibly should have no noticeable effect on anything), with the source thread, not the target, being partially responsible for checking whether the target is in kernel code: call SuspendThread, then GetThreadContext to synchronize with the suspended thread, and check whether it's sleeping in kernel code currently — if yes, notify it that suspension was requested and unsuspend it on the host, to let it self-suspend on the next kernel to guest return.